// grunt-insert-timestamp/Gruntfile.js
/*
 * grunt-insert-timestamp
 * https://github.com/cr0ybot/grunt-insert-timestamp
 *
 * Copyright (c) 2015 Cory Hughart
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    jshint: {
      all: [
        'Gruntfile.js',
        'tasks/*.js',
        '<%= nodeunit.tests %>'
      ],
      options: {
        jshintrc: '.jshintrc'
      }
    },

    // Before generating any new files, remove any previously-created files.
    clean: {
      tests: ['tmp']
    },

    // Configuration to be run (and then tested).
    insert_timestamp: {
      default_options: {
        options: {
          // test must define specific time and UTC format, otherwise untestable
          format: 'UTC:yyyy-mm-dd HH:MM:ss o',
          datetime: new Date('2015-03-14T09:26:53Z')
        },
        files: {
          'tmp/default_options': ['test/fixtures/testing', 'test/fixtures/123']
        }
      },
      custom_options: {
        options: {
          append: true,
          format: 'UTC:yyyy/mm/dd HH:MM:ss',
          template: '\n/*! Compiled on: {timestamp} */\n',
          datetime: new Date(0),
          insertNewlines: false
        },
        files: {
          'tmp/custom_options': ['test/fixtures/testing', 'test/fixtures/123']
        }
      }
    },

    // Unit tests.
    nodeunit: {
      tests: ['test/*_test.js']
    }

  });

  // Actually load this plugin's task(s).
  grunt.loadTasks('tasks');

  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-nodeunit');

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.
  grunt.registerTask('test', ['clean', 'insert_timestamp', 'nodeunit']);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['jshint', 'test']);

};
